File: test-pagination.py
import urllib.request
import json
from urllib.parse import urlencode

def test_pagination():
    has_more = True
    cursor = None
    page = 1
    total_items = 0
    
    while has_more and page <= 500:
        params = {"limit": 5}
        if cursor:
            params["cursor"] = cursor
        
        url = f"http://localhost:3000/api/listings?{urlencode(params)}"
        print(f"\nFetching page {page}: {url}")
        
        req = urllib.request.Request(url)
        try:
            with urllib.request.urlopen(req) as response:
                data = json.loads(response.read())
                items = data.get("data", [])
                
                print(f"Received {len(items)} items.")
                if items:
                    print(f"First item ID: {items[0]['id']}")
                    print(f"Last item ID: {items[-1]['id']}")
                
                total_items += len(items)
                pagination = data.get("pagination", {})
                has_more = pagination.get("has_more", False)
                cursor = pagination.get("next_cursor")
                
                print(f"Has More: {has_more}, Next Cursor: {'Present' if cursor else 'None'}")
                page += 1
                
        except Exception as e:
            print(f"Error: {e}")
            break
            
    print(f"\nTest finished. Fetched {total_items} items across {page - 1} pages.")

if __name__ == "__main__":
    test_pagination()

File: test-pagination.js
const fetch = require('node-fetch'); // we can just use global fetch in Node 18+

async function testPagination() {
  let hasMore = true;
  let cursor = null;
  let page = 1;
  let totalItems = 0;

  while (hasMore && page <= 5) {
    const url = `http://localhost:3000/api/listings?limit=5${cursor ? `&cursor=${encodeURIComponent(cursor)}` : ''}`;
    console.log(`\nFetching page ${page}: ${url}`);
    
    try {
      const res = await fetch(url);
      if (!res.ok) {
        console.error(`Error: ${res.status} ${await res.text()}`);
        break;
      }
      
      const data = await res.json();
      console.log(`Received ${data.data?.length || 0} items.`);
      if (data.data?.length > 0) {
        console.log(`Last item ID: ${data.data[data.data.length - 1].id}`);
        console.log(`Last item Score: ${data.data[data.data.length - 1].popularity_score}`);
      }
      
      totalItems += data.data?.length || 0;
      hasMore = data.pagination?.has_more;
      cursor = data.pagination?.next_cursor;
      
      console.log(`Has More: ${hasMore}, Next Cursor: ${cursor ? 'Present' : 'None'}`);
      page++;
    } catch (e) {
      console.error(e);
      break;
    }
  }
  
  console.log(`\nTest finished. Fetched ${totalItems} items across ${page - 1} pages.`);
}

testPagination();
